src/Http/Http.Extensions/src/RequestDelegateFactory.cs

```diff
@@ -256,16 +263,32 @@ private static Expression CreateArgument(ParameterInfo parameter, FactoryContext
             }
             else
             {
+
+                var nullability = nullabilityContext.Create(parameter);
+                var isOptional = parameter.HasDefaultValue || nullability.ReadState == NullabilityState.Nullable;
                 if (factoryContext.ServiceProviderIsService is IServiceProviderIsService serviceProviderIsService)
                 {
-                    // If the parameter resolves as a service then get it from services
-                    if (serviceProviderIsService.IsService(parameter.ParameterType))
+                    // If the parameter is required
+                    if (!isOptional)
                     {
-                        return Expression.Call(GetRequiredServiceMethod.MakeGenericMethod(parameter.ParameterType), RequestServicesExpr);
+                        // And we are able to resolve a service for it
+                        return serviceProviderIsService.IsService(parameter.ParameterType)
+                            ? Expression.Call(GetRequiredServiceMethod.MakeGenericMethod(parameter.ParameterType), RequestServicesExpr) // Then get it from the DI
+                            : BindParameterFromBody(parameter, allowEmpty: false, factoryContext); // Otherwise try to find it in the body
+                    }
+                    // If the parameter is optional
+                    else
+                    {
+                        // Then try to resolve it as an optional service and fallback to a body otherwise
+                        // Note: if the parameter provides a default value that value will be parsed
+                        // as part of the body, not as the service instance
+                        return Expression.Coalesce(
+                            Expression.Call(GetServiceMethod.MakeGenericMethod(parameter.ParameterType), RequestServicesExpr),
+                            BindParameterFromBody(parameter, allowEmpty: false, factoryContext));
```

Comment:
Is this falling back to the body when `IServiceProvider.GetService()` returns null during the request? If `IsService.IsService()` returns false for a given Type at build, we should always bind from the body if the parameter is optional or not.

We should probably log an error and return a 500 response in this case, but even passing through null is better than falling back to the body

Reply:
I was confused for a second. Since this is an optional case, I think passing in null rather than falling back to the body is the best behavior. It is weird for `IServiceProviderIsService.IsService()` return true for a type and give null later, but whatever.

Reply:
>  It is weird for IServiceProviderIsService.IsService() return true for a type and give null later, but whatever.

I'm not even sure how it could happen tbh.

Reply:
>  If IsService.IsService() returns false for a given Type at build, we should always bind from the body if the parameter is optional or not.

Why? This would make it impossible to use a service that was registered after the endpoint was defined and force users to order their dependencies before the route definitions. Maybe that makes sense but it feels like a tricky constraint to enforce.

Table below enumerates the current behavior.

| Parameter Optionality | IServiceProviderIsService.IsService() at build | GetService at runtime | Result |
|--|--|--|--|
| Required | True | True | Returns the service |
| Required | True | Null | 500s |
| Required | False | True | Returns the body |
| Required | False | Null | Returns the body |
| Optional | True | True | Returns the service |
| Optional | True | Null | Returns the body |
| Optional | False | True | Returns the service |
| Optional | False | Null | Returns the body |

Reply:
> This would make it impossible to use a service that was registered after the endpoint was defined and force users to order their dependencies before the route definitions (unless they use the FromService attribute).

This is not possible with the default ServiceProvider. If someone wants to use a custom container that cannot determine if the service exists at this point, they should use `[FromServices]` imo to make the intention clear.

Whether a parameter comes from a service or from the body should be determined at build time, not request time, period.

Reply:
@captainsafia and it doesn't matter that it can't happen with the default service provider. The code doesn't need to change, we were just talking about how it would work in practice.